src/centrality.rs

```diff
@@ -132,3 +135,127 @@ pub fn digraph_betweenness_centrality(
             .collect(),
     }
 }
+
+/// Compute the eigenvector centrality of a :class:`~PyGraph`.
+///
+/// For details on the eigenvector centrality refer to:
+///
+/// Phillip Bonacich. “Power and Centrality: A Family of Measures.”
+/// American Journal of Sociology 92(5):1170–1182, 1986
+/// <https://doi.org/10.1086/228631>
+///
+/// This function uses a power iteration method to compute the eigenvector
+/// and convergence is not guaranteed. The function will stop when `max_iter`
+/// iterations is reached or when the computed vector between two iterations
+/// is smaller than the error tolerance multiplied by the number of nodes.
+/// The implementation of this algorithm is based on the NetworkX
+/// `eigenvector_centrality() <https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.centrality.eigenvector_centrality.html>`__
+/// function.
+///
+/// In the case of multigraphs the weights of any parallel edges will be
+/// summed when computing the eigenvector centrality.
+///
+/// :param PyDigraph graph: The graph object to run the algorithm on
+/// :param weight_fn: An optional input callable that will be pased the edge's
+///     payload object and is expected to return a `float` weight for that edge.
+///     If this is not specified ``default_weight`` will be used as the weight
+///     for every edge in ``graph``
+/// :param float default_weight: If ``weight_fn`` is not set the default weight
+///     value to use for the weight of all edges
+/// :param int max_iter: The maximum number of iterations in the power method. If
+///     not specified a default value of 100 is used.
+/// :param float tol: The error tolerance used when checking for convergence in the
+///     power method. If this is not specified default value of 1e-6 is used.
+///
+/// :returns: a read-only dict-like object whose keys are the node indices and values are the
+///      centrality score for that node.
+/// :rtype: CentralityMapping
+#[pyfunction(default_weight = "1.0", max_iter = "100", tol = "1e-6")]
+#[pyo3(text_signature = "(graph, /, weight_fn=None, default_weight=1.0, max_iter=100, tol=1e-6)")]
+pub fn graph_eigenvector_centrality(
+    py: Python,
+    graph: &graph::PyGraph,
+    weight_fn: Option<PyObject>,
+    default_weight: f64,
+    max_iter: usize,
+    tol: f64,
+) -> PyResult<CentralityMapping> {
+    let cost_fn = CostFn::try_from((weight_fn, default_weight))?;
+    let ev_centrality = centrality::eigenvector_centrality(
+        &graph.graph,
+        |e| cost_fn.call(py, e.weight()),
```

Comment:
It might be beneficial to precompute and store all edge weights like we do in `all_pairs_dijkstra_path_lengths` so we avoid calling Python overhead each iteration.

Reply:
Let me benchmark this to see if it makes a difference

Reply:
I benchmarked this and it's basically the same but marginally faster in a couple of cases:

![eigv_centrality_times_py_callback](https://user-images.githubusercontent.com/2447371/172892209-97c1169e-48cb-4031-b4c4-b3e8e66e30a2.png)

not really a noticeable difference. But I think I'll make the change anyway, mostly because it does open up potential parallelization. I think I can make the inner loop execute in parallel and will test that in bit.

Reply:
I pushed this up in: https://github.com/Qiskit/retworkx/pull/621/commits/3204e09eb55509860ead214a116755ec86148637 I'm still waiting on the data collection for the run with a python callback. It's been running the sweep for ~30min and the precomputed weights sweep took ~15min. So there is no doubt this is faster. I'll generate the graph when it eventually finishes just for posterity.

Reply:
Is the graph with a callback set (I just used `float`).

![eigv_centrality_times_py_callback](https://user-images.githubusercontent.com/2447371/172910852-068b01d4-b6a8-49a7-b483-bbbbdda3a372.png)